storage/aptosdb/src/pruner/worker.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) The Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+use aptos_types::transaction::Version;
+use schemadb::DB;
+
+use crate::pruner::{
+    db_pruner::DBPruner,
+    transaction_store::TransactionStorePruner,
+};
+use aptos_infallible::Mutex;
+
+use std::{
+    sync::{mpsc::Receiver, Arc},
+    time::Instant,
+};
+use crate::pruner::state_store::StateStorePruner;
+
+/// Maintains all the DBPruners and periodically calls the db_pruner's prune method to prune the DB.
+/// This also exposes API to report the progress to the parent thread.
+pub struct Worker {
+    command_receiver: Receiver<Command>,
+    /// Keeps tracks of all the DB pruners
+    db_pruners: Vec<Mutex<Arc<dyn DBPruner + Send + Sync>>>,
+    /// Keeps a record of the pruning progress. If this equals to version `V`, we know versions
+    /// smaller than `V` are no longer readable.
+    /// This being an atomic value is to communicate the info with the Pruner thread (for tests).
+    least_readable_versions: Arc<Mutex<Vec<Version>>>,
+    /// Indicates if there's NOT any pending work to do currently, to hint
+    /// `Self::receive_commands()` to `recv()` blocking-ly.
+    blocking_recv: bool,
+}
+
+impl Worker {
+    const MAX_VERSIONS_TO_PRUNE_PER_BATCH: usize = 100;
+
+    pub(crate) fn new(
+        db: Arc<DB>,
+        command_receiver: Receiver<Command>,
+        least_readable_versions: Arc<Mutex<Vec<Version>>>,
+    ) -> Self {
+        Self {
+            db_pruners: vec![
+                Mutex::new(Arc::new(StateStorePruner::new(
+                    Arc::clone(&db),
+                    0,
+                    Instant::now(),
+                ))),
+                Mutex::new(Arc::new(TransactionStorePruner::new(Arc::clone(&db)))),
+            ],
+            command_receiver,
+            least_readable_versions,
+            blocking_recv: true,
+        }
+    }
+
+    pub(crate) fn work(mut self) {
+        for db_pruner in &self.db_pruners {
+            db_pruner.lock().initialize();
+        }
+        while self.receive_commands() {
+            // Process a reasonably small batch of work before trying to receive commands again,
+            // in case `Command::Quit` is received (that's when we should quit.)
+            for db_pruner in &self.db_pruners {
```

Comment:
@msmouse - I see, thanks for the explanation. This will actually be messy because going forward, I am thinking of implementing a change to keep all the pruners in sync and commit the changes atomically across pruners. If we keep changing the version across pruners, then achieving atomic commits across pruners can be tricky. Do you have any thoughts?

Reply:
That's fine, guess we can just make each of the pruners yield a smaller change set.

Reply:
Let me make the max_version to prune a config knob for this. WDYT?

Reply:
Yeah, making that tunable would be nice.

Reply:
I will do that in a subsequent PR.